### PR TITLE
Fungible tokens implementation

### PR DESCRIPTION
**Pull request by liran-funaro**
Signed-off-by: Liran Funaro <liran.funaro@gmail.com>

### CODE REVIEW

File: internal/tokens/users.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+import (
+	"github.com/copa-europe-tokens/internal/common"
+	"github.com/hyperledger-labs/orion-sdk-go/pkg/bcdb"
+	oriontypes "github.com/hyperledger-labs/orion-server/pkg/types"
+	"github.com/pkg/errors"
+)
+
+type UserContext struct {
+	m      *Manager
```
Comment (tock-ibm) [inline internal/tokens/users.go:11]: same comment as above, have a "UserOperator" that the manager creates and owns, and has what is needed.

File: pkg/server/server_test.go [hunk 2]
```diff
@@ -140,6 +192,40 @@ func TestTokensServer_MainFlow(t *testing.T) {
...
+	_, bobKeyPath := c.GetUserCertKeyPath("bob")
+	hashSignerBob, err := tokenscrypto.NewSigner("bob", bobKeyPath)
+
+	certCharlie, signerCharlie := testutils.LoadTestCrypto(t, c.GetUserCertDir(), "charlie")
+	_, charlieKeyPath := c.GetUserCertKeyPath("charlie")
+	hashSignerCharlie, err := tokenscrypto.NewSigner("charlie", charlieKeyPath)
+
+	// Add 2 users
+	// The test environment prepares crypto material for: server, admin, alice, bob, and charlie; alice is the custodian.
+
```
Comment (tock-ibm) [inline pkg/server/server_test.go:210]: we can assume that the initialization sequence coming from the market place is as following: 
- deploy NFT
- deploy annotations
- create reserve owner
- deploy fungible
- add users

This simplifies the treatment of transfer in fungible, as if the target does not have an account, you do not have to auto-create it.

File: pkg/server/server_test.go [hunk 3]
```diff
@@ -394,12 +450,164 @@ func TestTokensServer_MainFlow(t *testing.T) {
...
 
+	t.Run("Fungible", func(t *testing.T) {
+		fungibleResponse := types.FungibleDeployResponse{}
+		env.testPostRequest(t, "deploy",
+			constants.FungibleDeploy,
+			&types.FungibleDeployRequest{
+				Name:         "Fungible test",
+				ReserveOwner: "bob",
+			},
+			&fungibleResponse,
```
Comment (tock-ibm) [inline pkg/server/server_test.go:461]: Here, and in all the tests below, you have to check that the response content is what you expect it to be.

File: pkg/server/server.go [hunk 4]
```diff
@@ -42,6 +42,7 @@ func NewTokensServer(conf *config.Configuration, lg *logger.SugarLogger) (*Token
 	mux.Handle(constants.TokensUsersSubTree, httphandlers.NewUserHandler(tokenManager, lg))
 	mux.Handle(constants.TokensAnnotationsEndpoint, httphandlers.NewAnnotationsHandler(tokenManager, lg))
 	mux.Handle(constants.TokensAnnotationsSubTree, httphandlers.NewAnnotationsHandler(tokenManager, lg))
+	mux.Handle(constants.FungibleEndpoint, httphandlers.NewFungibleHandler(tokenManager, lg))
```
Comment (tock-ibm) [inline pkg/server/server.go:45]: How come you don't need the subtree?

File: internal/tokens/manager.go [hunk 5]
```diff
@@ -89,12 +90,25 @@ type Manager struct {
...
 }
 
-type TokenDescription struct {
+func abort(ctx bcdb.TxContext) {
+	if ctx != nil {
+		_ = ctx.Abort()
+	}
+}
+
+type CommonTokenDescription struct {
```
Comment (tock-ibm) [inline internal/tokens/manager.go:99]: The value stored in the typeDB could be an internal struct that contains all of these and a 5th field "Extension" that contains a class dependent object (string or object), which is interpreted according to Class.

File: pkg/server/server.go (hunk 4 above)
Comment (liran-funaro) [reply pkg/server/server.go:45]: Because there isn't an API call at /tokens/fungible

File: internal/tokens/manager_test.go [hunk 6]
```diff
@@ -1320,3 +1341,865 @@ func (e *testEnv) clean() {
...
+			Owner:    "nonuser",
+			NewOwner: "bob",
+			Quantity: 1,
+		}
+		response, err := env.manager.FungiblePrepareTransfer(typeId, transferRequest)
+		assertTokenHttpErr(t, http.StatusNotFound, response, err)
+	})
+
+	t.Run("error: new owner does not exists", func(t *testing.T) {
+		transferRequest := &types.FungibleTransferRequest{
```
Comment (tock-ibm) [inline internal/tokens/manager_test.go:1906]: add use case to verify that one cannot transfer to self

File: internal/tokens/manager_test.go [hunk 7]
```diff
@@ -1320,3 +1341,865 @@ func (e *testEnv) clean() {
...
+		assertRecordEqual(t, types.FungibleAccountRecord{
+			Account: mainAccount,
+			Owner:   "dave",
+			Balance: 3,
+			Comment: mainAccount,
+		}, accList[0])
+
+		accList, err = env.manager.FungibleAccounts(typeId, "dave", "")
+		require.NoError(t, err)
+		assert.Equal(t, 3, len(accList))
```
Comment (tock-ibm) [inline internal/tokens/manager_test.go:2101]: make sure the remaining accounts are who you think they are

File: internal/tokens/manager_test.go [hunk 8]
```diff
@@ -1320,3 +1341,865 @@ func (e *testEnv) clean() {
...
+		require.NoError(t, err)
+		assert.Equal(t, 2, len(accList))
+	})
+
+	t.Run("error: no accounts to consolidate", func(t *testing.T) {
+		request := &types.FungibleConsolidateRequest{
+			Owner: "bob",
+		}
+		response, err := env.manager.FungiblePrepareConsolidate(typeId, request)
+		assertTokenHttpErr(t, http.StatusNotFound, response, err)
```
Comment (tock-ibm) [inline internal/tokens/manager_test.go:2137]: again, match the error message as well...

File: internal/tokens/token.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+// ResetTx creates a new transaction. It will abort previous transaction if existed.
+func (ctx *TokenTxContext) ResetTx() error {
+	if ctx.dataTx != nil {
+		if err := ctx.dataTx.Abort(); err != nil {
+			return err
+		}
+	}
+
+	dataTx, err := ctx.custodianSession.DataTx()
```
Comment (tock-ibm) [inline internal/tokens/token.go:64]: Note that here, a txID is computed. If you want to use an externally generated TxID, use WithTxID()

File: internal/httphandlers/utils.go [hunk 10]
```diff
@@ -46,7 +46,7 @@ func decode(request *http.Request, requestBody interface{}) error {
 	dec := json.NewDecoder(request.Body)
 	dec.DisallowUnknownFields()
 	if err := dec.Decode(&requestBody); err != nil {
-		return common.WrapErrInvalid(err)
+		return common.NewErrInvalid("Request does not follow the required format: %s", err)
```
Comment (tock-ibm) [inline internal/httphandlers/utils.go:49]: It is a convention in golang that error messages always start with lower case letter.

**Review by tock-ibm — COMMENTED** [review]
This too much code to review at once. I hope I got it all. Netx time lets create different PRs for implementation steps, eg.: deploy, describe, transfer, consolidate, etc.

File: internal/tokens/manager_test.go [hunk 11]
```diff
@@ -1320,3 +1341,865 @@ func (e *testEnv) clean() {
...
+	return assert.Equal(t, expectedStatus, tknErr.StatusCode, "Error: %v", actualErr)
+}
+
+func assertTokenHttpErrMessage(t *testing.T, expectedStatus int, expectedMessage string, actualResponse interface{}, actualErr error) bool {
+	if !assertTokenHttpErr(t, expectedStatus, actualResponse, actualErr) {
+		return false
+	}
+
+	return assert.EqualError(t, actualErr, expectedMessage)
+}
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply internal/tokens/manager_test.go:1372]: Testing the wording of error messages is not a good practice IMO. Any change in the wording of the messages will break the tests. If you think it is essential to differentiate between "user not found", "account not found", etc., then I suggest adding more elaborated error codes and testing those.

File: internal/tokens/manager.go [hunk 12]
```diff
@@ -1176,34 +1199,264 @@ func (m *Manager) createTypesDB() (err error) {
...
+
+	reserve, err := ctx.getReserveAccount()
+	if err != nil {
+		return nil, err
+	}
+
+	reserveOwner, err := ctx.getReserveOwner()
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply internal/tokens/manager.go:1278]: The description is cached in the `ctx` object, so this does not issue another GET.

File: internal/tokens/manager.go [hunk 13]
```diff
@@ -1176,34 +1199,264 @@ func (m *Manager) createTypesDB() (err error) {
...
+
+	// Retry until we got a response.
+	var response *types.FungibleTransferResponse = nil
+	for response == nil {
+		response, err = ctx.internalFungiblePrepareTransfer(request)
+		if err != nil {
+			return nil, err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply internal/tokens/manager.go:1351]: https://github.com/tock-ibm/copa-europe-tokens/issues/35

File: internal/httphandlers/utils.go (hunk 10 above)
Comment (liran-funaro) [reply internal/httphandlers/utils.go:49]: These error messages are reported to the user via HTTP, so these are actually HTTP error messages and as such, they are expected to be grammatically correct. How do you suggest reconciling this?

**Review by tock-ibm — APPROVED** [review]

**Pull request merged**